corehq/apps/userreports/reports/builder/forms.py

```diff
@@ -0,0 +1,569 @@
+            ('table', _("Table")),
+        ],
+    )
+
+    application = forms.ChoiceField()
+    source_type = forms.ChoiceField(choices=[
+        ("case", _("Case")),
+        ("form", _("Form")),
+    ])
+    report_source = forms.ChoiceField()
+
+    def __init__(self, domain, *args, **kwargs):
+        super(CreateNewReportForm, self).__init__(*args, **kwargs)
+
+        self.domain = domain
+        apps = get_apps_in_domain(domain, full=True, include_remote=False)
+        self.fields['application'].choices = [
+            (app._id, app.name) for app in apps
+        ]
+        self.fields['report_source'].choices = [
+            (ct, ct) for app in apps for ct in app.get_case_types()
+        ] + [
+            (form_id, form_id) for form_id in [
+                form.get_unique_id() for app in apps for form in app.get_forms()
+            ]
+        ]
+
+        self.helper = FormHelper()
+        self.helper.form_class = "form-horizontal"
+        self.helper.form_id = "report-builder-form"
+        self.helper.layout = crispy.Layout(
+            crispy.Fieldset(
+                _('Create New Report'),
+                'report_type',
+                crispy.Field('application', data_bind='value: application'),
+                crispy.Field('source_type', data_bind='value: sourceType'),
+                crispy.Field('report_source', data_bind='''
+                    options: sourcesMap[application()][sourceType()],
+                    optionsText: function(item){return item.text},
+                    optionsValue: function(item){return item.value}
+                '''),
+            ),
+            FormActions(
+                crispy.ButtonHolder(
+                    crispy.Submit(
+                        'create_new_report_builder_btn',
+                        _('Next'),
+                    )
+                ),
+            ),
+        )
+
+    def clean(self):
+        """
+        Raise a validation error if there are already 5 data sources and this
+        report won't be able to use one of the existing ones.
+        """
+        cleaned_data = super(CreateNewReportForm, self).clean()
+        source_type = cleaned_data.get('source_type')
+        report_source = cleaned_data.get('report_source')
+        app_id = cleaned_data.get('application')
+
+        if report_source and source_type and app_id:
+            app = Application.get(app_id)
+            ds_builder = DataSourceBuilder(self.domain, app, source_type, report_source)
+
+            existing_sources = DataSourceConfiguration.by_domain(self.domain)
+            if len(existing_sources) >= 5:
+                suitable_source_exists = False
+                for s in existing_sources:
+                    if (
+                        s.referenced_doc_type == ds_builder.source_doc_type and
+                        s.configured_filter == ds_builder.filter and
+                        s.configured_indicators == ds_builder.indicators
```

Comment:
Updated in 5298395 to use app + xmlns/case type + app build number. There's a potentially big downside here though in that any app change will "invalidate" the data source, even if it doesn't affect the case type or form that the data source is based on. But, perhaps this is satisfactory until a more comprehensive solution is found (such as calculating diffs between needed data sources and existing data sources or something like that).